io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shelly3em/IoShelly3EmImpl.java

```diff
@@ -0,0 +1,283 @@
+)
+@EventTopics({ //
+		EdgeEventConstants.TOPIC_CYCLE_AFTER_PROCESS_IMAGE, //
+		EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE //
+})
+public class IoShelly3EmImpl extends AbstractOpenemsComponent
+		implements IoShelly3Em, DigitalOutput, ElectricityMeter, OpenemsComponent, TimedataProvider, EventHandler {
+
+	private final CalculateEnergyFromPower calculateProductionEnergy = new CalculateEnergyFromPower(this,
+			ElectricityMeter.ChannelId.ACTIVE_PRODUCTION_ENERGY);
+	private final CalculateEnergyFromPower calculateConsumptionEnergy = new CalculateEnergyFromPower(this,
+			ElectricityMeter.ChannelId.ACTIVE_CONSUMPTION_ENERGY);
+
+	private final Logger log = LoggerFactory.getLogger(IoShelly3EmImpl.class);
+	private final BooleanWriteChannel[] digitalOutputChannels;
+
+	private MeterType meterType = null;
+	private String baseUrl;
+
+	@Reference(policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY, cardinality = ReferenceCardinality.OPTIONAL)
+	private volatile Timedata timedata;
+
+	@Reference()
+	private BridgeHttpFactory httpBridgeFactory;
+	private BridgeHttp httpBridge;
+
+	public IoShelly3EmImpl() {
+		super(//
+				OpenemsComponent.ChannelId.values(), //
+				ElectricityMeter.ChannelId.values(), //
+				DigitalOutput.ChannelId.values(), //
+				IoShelly3Em.ChannelId.values() //
+		);
+		this.digitalOutputChannels = new BooleanWriteChannel[] { this.channel(IoShelly3Em.ChannelId.RELAY) };
+
+		ElectricityMeter.calculateSumActivePowerFromPhases(this);
+	}
+
+	@Activate
+	protected void activate(ComponentContext context, Config config) {
+		super.activate(context, config.id(), config.alias(), config.enabled());
+		this.meterType = config.type();
+		this.baseUrl = "http://" + config.ip();
+		this.httpBridge = this.httpBridgeFactory.get();
+
+		if (this.isEnabled()) {
+			this.httpBridge.subscribeJsonEveryCycle(this.baseUrl + "/status", this::processHttpResult);
+		}
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		this.httpBridgeFactory.unget(this.httpBridge);
+		this.httpBridge = null;
+		super.deactivate();
+	}
+
+	@Override
+	public BooleanWriteChannel[] digitalOutputChannels() {
+		return this.digitalOutputChannels;
+	}
+
+	@Override
+	public String debugLog() {
+		var b = new StringBuilder();
+		var valueOpt = this.getRelayChannel().value().asOptional();
+		b.append(valueOpt.isPresent() //
+				? (valueOpt.get() //
+						? "ON" //
+						: "OFF") //
+				: "Unknown");
+		b.append("|");
+		b.append(this.getActivePowerChannel().value().asString());
+
+		return b.toString();
+	}
+
+	@Override
+	public void handleEvent(Event event) {
+		if (!this.isEnabled()) {
+			return;
+		}
+
+		switch (event.getTopic()) {
+		case EdgeEventConstants.TOPIC_CYCLE_AFTER_PROCESS_IMAGE //
+			-> this.calculateEnergy();
+		case EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE //
+			-> this.executeWrite();
+		}
+	}
+
+	private void processHttpResult(JsonElement result, Throwable error) {
```

Comment:
For future PRs, please provide an example Json result. Then we can write proper JUnit tests. Right now this code has only 33 % code coverage.

Reply:
Example JSON:

{"wifi_sta":{"connected":true,"ssid":"Fritzlbox","ip":"192.168.178.130","rssi":-41},"cloud":{"enabled":true,"connected":true},"mqtt":{"connected":false},"time":"11:31","unixtime":1708770669,"serial":5,"has_update":false,"mac":"485519D6A289","cfg_changed_cnt":0,"actions_stats":{"skipped":0},"relays":[{"ison":false,"has_timer":false,"timer_started":0,"timer_duration":0,"timer_remaining":0,"overpower":false,"is_valid":true,"source":"http"}],"emeters":[{"power":3.81,"pf":0.33,"current":0.05,"voltage":231.38,"is_valid":true,"total":0.6,"total_returned":0.0},{"power":3.86,"pf":0.33,"current":0.05,"voltage":231.36,"is_valid":true,"total":0.6,"total_returned":0.0},{"power":3.83,"pf":0.33,"current":0.05,"voltage":231.36,"is_valid":true,"total":0.6,"total_returned":0.0}],"total_power":11.50,"emeter_n":{"current":0.00,"ixsum":0.15,"mismatch":false,"is_valid":false},"fs_mounted":true,"v_data":1,"ct_calst":0,"update":{"status":"idle","has_update":false,"new_version":"20230913-114244/v1.14.0-gcb84623","old_version":"20230913-114244/v1.14.0-gcb84623","beta_version":"20231107-165007/v1.14.1-rc1-g0617c15"},"ram_total":49920,"ram_free":29108,"fs_size":233681,"fs_free":155118,"uptime":251}